integration-tests/test_events_decode/lib.rs

```diff
@@ -0,0 +1,241 @@
+//! This is a reference implementation with one approach to decoding
+//! (capturing emitted) events within unit and E2E tests.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[ink::contract]
+mod testing_event_decode {
+
+    #[ink(storage)]
+    pub struct TestingEventDecode {}
+
+    #[ink(event)]
+    pub struct FlipEvent {
+        #[ink(topic)]
+        flipper: AccountId,
+        #[ink(topic)]
+        value: bool,
+    }
+
+    #[ink(event)]
+    pub struct FlopEvent {
+        #[ink(topic)]
+        flipper: AccountId,
+        #[ink(topic)]
+        flopper: AccountId,
+        #[ink(topic)]
+        value: bool,
+    }
+
+    impl Default for TestingEventDecode {
+        fn default() -> Self {
+            Self::new()
+        }
+    }
+
+    impl TestingEventDecode {
+
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            Self {}
+        }
+
+        /// This message emits two events for test verification.
+        #[ink(message)]
+        pub fn flipflop(&mut self) {
+            let caller = self.env().caller();
+            let contract = self.env().account_id();
+
+            // Emit FlipEvent.
+            self.env().emit_event(FlipEvent {
+                flipper: caller,
+                value: true,
+            });
+
+            // Emit FlopEvent.
+            self.env().emit_event(FlopEvent {
+                flipper: caller,
+                flopper: contract,
+                value: false,
+            });
+        }
+    }
+
+    /// This is one way to capture and check event emission in ink unit tests.
+    #[cfg(test)]
+    mod tests {
+
+        use super::*;
+        use ink::env::test::EmittedEvent;
+
+        type Event = <TestingEventDecode as ::ink::reflect::ContractEventBase>::Type;
+
+        fn decode_events(emitted_events: Vec<EmittedEvent>) -> Vec<Event> {
```

Comment:
This is actually a very useful function to be added to e2e create directly. Can you please look into this?

Reply:
To clarify, do you mean reimplement this `decode_events` function to work within the ink_e2e test scope? Or do you mean include this particular `decode_events` function as `ink::env::test::decode_events`? Or both?

If the former, this would be part of the followup issue regarding improving event decode to not be so awkward.

In any case, this might be something @ganesh1233456 or @alessandro-baldassarre may wish to address individually. I will see to it that the proper issues are created.

Reply:
I think for unit tests, we want to have something like `ink::env::test::recorded_events().decode()` or `ink::env::test::decoded_events()`.  To check whether the events have been emitted, you can use something like  `.contains()` or similar list functions.

With e2e tests it's a little bit trickier, we probably want to have something like `let flip_event = flipflop_result.events.find_decoded::<FlipEvent>()` or similar to automatically check whether the event has been emitted, decode it and then assert on its values.